File: hr.js
// Enhanced HR Dashboard JavaScript
document.addEventListener('DOMContentLoaded', function() {
    // Initialize with dashboard data
    loadDashboardData();
    showDashboard();
    
    // Add event listeners
    setupEventListeners();
  });
  
  // Configuration
  const CONFIG = {
    roleSalaries: {
      'Accountant': 15000,
      'Manager': 30000,
      'Security': 5000,
      'Cleaner': 5000,
      'Finance': 10000,
      'HR': 12000
    },
    apiBaseUrl: 'http://localhost:8000/api'
  };
  
  // Set up all event listeners
  function setupEventListeners() {
    // Role salary autofill
    document.getElementById('role').addEventListener('change', function() {
      const selectedRole = this.value;
      const salaryInput = document.getElementById('salary');
      salaryInput.value = CONFIG.roleSalaries[selectedRole] || '';
      if (salaryInput.value) {
        salaryInput.classList.add('success-glow');
        setTimeout(() => salaryInput.classList.remove('success-glow'), 1000);
      }
    });
  
    // Form submissions
    document.getElementById('hiringForm').addEventListener('submit', handleHireEmployee);
    document.getElementById('firingForm').addEventListener('submit', handleFireEmployee);
    
    // Modal interactions
    document.querySelector('.close-modal').addEventListener('click', closeModal);
    document.getElementById('overlay').addEventListener('click', toggleSidebar);
    
    // Sidebar buttons
    document.querySelector('.sidebar button:not(.logout-btn)').addEventListener('click', openModal);
    document.querySelector('.logout-btn').addEventListener('click', logout);
  }
  
  // Page Navigation Functions
  function showDashboard() {
    document.getElementById('hrDashboard').style.display = 'flex';
    document.getElementById('hirePage').style.display = 'none';
    document.getElementById('firePage').style.display = 'none';
    loadDashboardData();
  }
  
  function showHirePage() {
    document.getElementById('hrDashboard').style.display = 'none';
    document.getElementById('hirePage').style.display = 'flex';
    document.getElementById('firePage').style.display = 'none';
    document.getElementById('hire-name').focus();
  }
  
  function showFirePage() {
    document.getElementById('hrDashboard').style.display = 'none';
    document.getElementById('hirePage').style.display = 'none';
    document.getElementById('firePage').style.display = 'flex';
    document.getElementById('fire-name').focus();
  }
  
  // Sidebar and Modal Functions
  function toggleSidebar() {
    const sidebar = document.getElementById('sidebar');
    const overlay = document.getElementById('overlay');
    const hamburger = document.querySelector('.hamburger-menu');
  
    sidebar.classList.toggle('open');
    overlay.classList.toggle('active');
    hamburger.classList.toggle('open');
    
    // Add animation class
    if (sidebar.classList.contains('open')) {
      sidebar.classList.add('sidebar-open-animation');
      setTimeout(() => sidebar.classList.remove('sidebar-open-animation'), 500);
    }
  }
  
  function openModal() {
    document.getElementById('modal').style.display = 'flex';
    document.getElementById('currentUsername').focus();
    toggleSidebar();
  }
  
  function closeModal() {
    document.getElementById('modal').style.display = 'none';
    document.getElementById('credentialsForm').reset();
  }
  
  // Data Loading Functions
  async function loadDashboardData() {
    try {
      await Promise.all([
        loadHRManagerData(),
        loadDepartmentStats(),
        loadRecentActions()
      ]);
      
      // Add subtle animation when data loads
      document.querySelector('.dashboard-container').classList.add('data-loaded');
      setTimeout(() => {
        document.querySelector('.dashboard-container').classList.remove('data-loaded');
      }, 1000);
    } catch (error) {
      console.error('Error loading dashboard data:', error);
      showToast('Failed to load dashboard data', 'error');
    }
  }
  
  async function loadHRManagerData() {
    try {
      const response = await fetch(`${CONFIG.apiBaseUrl}/current-user`);
      const result = await response.json();
      
      if (!result.success) throw new Error(result.error || 'Failed to load user data');
      
      document.getElementById('userName').textContent = result.user.name;
      document.getElementById('userEmpId').textContent = result.user.emp_id;
      document.getElementById('userRole').textContent = result.user.role;
      document.getElementById('hrName').textContent = result.user.name;
      
      // Store in sessionStorage
      sessionStorage.setItem('hrName', result.user.name);
      sessionStorage.setItem('hrEmpId', result.user.emp_id);
      sessionStorage.setItem('hrRole', result.user.role);
      
    } catch (error) {
      console.error('Error loading HR data:', error);
      // Fallback to session storage
      document.getElementById('userName').textContent = sessionStorage.getItem('hrName') || 'HR Manager';
      document.getElementById('userEmpId').textContent = sessionStorage.getItem('hrEmpId') || 'HR001';
      document.getElementById('userRole').textContent = sessionStorage.getItem('hrRole') || 'HR Manager';
      document.getElementById('hrName').textContent = sessionStorage.getItem('hrName') || 'HR Manager';
    }
  }
  
  async function loadDepartmentStats() {
    try {
      const response = await fetch(`${CONFIG.apiBaseUrl}/department-stats`);
      const result = await response.json();
      
      if (!result.success) throw new Error(result.error || 'Failed to load stats');
      
      const statsHTML = result.stats.map(stat => `
        <div class="stat-item">
          <span class="stat-label">${stat.dep_name}:</span>
          <span class="stat-value">${stat.employee_count}</span>
        </div>
      `).join('');
      
      document.getElementById('statsContent').innerHTML = statsHTML;
      
    } catch (error) {
      console.error('Error loading department stats:', error);
      document.getElementById('statsContent').innerHTML = `
        <div class="error-message">
          <i class="fas fa-exclamation-circle"></i> Failed to load statistics
        </div>
      `;
    }
  }
  
  async function loadRecentActions() {
    try {
      const response = await fetch(`${CONFIG.apiBaseUrl}/recent-actions`);
      const result = await response.json();
      
      if (!result.success) throw new Error(result.error || 'Failed to load actions');
      
      const actionsHTML = result.actions.map(action => `
        <div class="action-item">
          <div class="action-type">
            <i class="fas ${getActionIcon(action.action_type)}"></i>
            ${action.action_type}: ${action.emp_id}
          </div>
          <div>${action.details}</div>
          <div class="action-date">
            <i class="far fa-clock"></i> ${new Date(action.action_date).toLocaleString()}
          </div>
        </div>
      `).join('');
      
      document.getElementById('actionsContent').innerHTML = actionsHTML;
      
    } catch (error) {
      console.error('Error loading recent actions:', error);
      document.getElementById('actionsContent').innerHTML = `
        <div class="error-message">
          <i class="fas fa-exclamation-circle"></i> Failed to load recent actions
        </div>
      `;
    }
  }
  
  function getActionIcon(actionType) {
    const icons = {
      'Hire': 'fa-user-plus',
      'Fire': 'fa-user-minus',
      'Update': 'fa-user-edit',
      'Promotion': 'fa-arrow-up'
    };
    return icons[actionType] || 'fa-user-cog';
  }
  
  // Form Handlers
  async function handleHireEmployee(event) {
    event.preventDefault();
    const form = event.target;
    const submitBtn = form.querySelector('button[type="submit"]');
    
    try {
      // Disable button and show loading state
      submitBtn.disabled = true;
      submitBtn.innerHTML = '<i class="fas fa-spinner fa-spin"></i> Processing...';
      
      // Collect form data
      const formData = {
        name: document.getElementById('hire-name').value.trim(),
        phone: document.getElementById('phone').value.trim(),
        email: document.getElementById('email').value.trim(),
        city: document.getElementById('city').value.trim(),
        address: document.getElementById('address').value.trim(),
        salary: document.getElementById('salary').value.trim(),
        role: document.getElementById('role').value.trim(),
        branch_id: document.getElementById('branch').value.trim(),
        dob: document.getElementById('dob').value.trim(),
        gender: document.getElementById('gender').value.trim()
      };
  
      // Validate required fields
      if (Object.values(formData).some(value => !value)) {
        throw new Error('All fields are required');
      }
  
      // Send data to backend
      const response = await fetch(`${CONFIG.apiBaseUrl}/hire`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(formData)
      });
  
      const result = await response.json();
  
      if (!result.success) {
        throw new Error(result.error || 'Failed to hire employee');
      }
  
      // Show success popup
      showSuccessPopup(formData, result);
      
      // Reset form
      form.reset();
      
      // Update UI
      addNewActionToUI({
        action_type: 'Hire',
        emp_name: formData.name,
        details: `Hired as ${formData.role} with salary ${formData.salary}`,
        action_date: new Date().toISOString()
      });
      
      // Reload data
      loadDepartmentStats();
      
    } catch (error) {
      console.error('Hiring error:', error);
      showToast(`Error: ${error.message}`, 'error');
    } finally {
      submitBtn.disabled = false;
      submitBtn.textContent = 'Hire Employee';
    }
  }
  
  async function handleFireEmployee(event) {
    event.preventDefault();
    const form = event.target;
    const submitBtn = form.querySelector('button[type="submit"]');
    
    try {
      submitBtn.disabled = true;
      submitBtn.innerHTML = '<i class="fas fa-spinner fa-spin"></i> Processing...';
      
      const formData = {
        name: document.getElementById('fire-name').value.trim(),
        empId: document.getElementById('empId').value.trim()
      };
  
      if (!formData.name || !formData.empId) {
        throw new Error('Both name and employee ID are required');
      }
      
      const response = await fetch(`${CONFIG.apiBaseUrl}/fire`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(formData)
      });
  
      const result = await response.json();
  
      if (!result.success) {
        throw new Error(result.error || 'Failed to terminate employee');
      }
  
      showFirePopup(formData);
      form.reset();
      loadRecentActions();
      loadDepartmentStats();
      
    } catch (error) {
      console.error('Firing error:', error);
      showToast(`Error: ${error.message}`, 'error');
    } finally {
      submitBtn.disabled = false;
      submitBtn.textContent = 'Terminate Employee';
    }
  }
  
  async function updateCredentials() {
    const form = document.getElementById('credentialsForm');
    const currentUsername = document.getElementById('currentUsername').value;
    const newUsername = document.getElementById('newUsername').value;
    const currentPassword = document.getElementById('currentPassword').value;
    const newPassword = document.getElementById('newPassword').value;
  
    if (!currentUsername || !newUsername || !currentPassword || !newPassword) {
      showToast('Please fill in all fields', 'warning');
      return;
    }
  
    try {
      const response = await fetch(`${CONFIG.apiBaseUrl}/update-credentials`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          currentUsername,
          newUsername,
          currentPassword,
          newPassword
        })
      });
  
      const result = await response.json();
  
      if (!result.success) {
        throw new Error(result.error || 'Failed to update credentials');
      }
  
      showToast('Credentials updated successfully!', 'success');
      closeModal();
      form.reset();
      
    } catch (error) {
      console.error('Error updating credentials:', error);
      showToast(`Error: ${error.message}`, 'error');
    }
  }
  
  // UI Helper Functions
  function showSuccessPopup(formData, result) {
    const successHTML = `
      <div class="success-popup">
        <h3>🎉 Employee Hired Successfully! 🎉</h3>
        <div class="employee-card">
          <div class="emp-name">${formData.name}</div>
          <div class="emp-id">Employee ID: <span>${result.emp_id}</span></div>
          <div class="emp-credentials">
            <p><strong>Username:</strong> ${result.username}</p>
            <p><strong>Temporary Password:</strong> ${result.temp_password}</p>
          </div>
          <div class="warning">⚠️ Employee must change password on first login</div>
        </div>
        <button onclick="closeSuccessPopup()">Close</button>
      </div>
    `;
    
    document.body.insertAdjacentHTML('beforeend', successHTML);
  }
  
  function showFirePopup(formData) {
    const fireHTML = `
      <div class="fire-popup">
        <div class="fire-animation">🔥</div>
        <h3>Employee Terminated</h3>
        <p>${formData.name} (ID: ${formData.empId}) has been terminated.</p>
        <p class="warning">Access revoked immediately!</p>
        <button onclick="closeFirePopup()">Acknowledge</button>
      </div>
    `;
    
    document.body.insertAdjacentHTML('beforeend', fireHTML);
  }
  
  function addNewActionToUI(action) {
    const actionsContainer = document.getElementById('actionsContent');
    const actionHTML = `
      <div class="action-item new-action">
        <div class="action-type">
          <i class="fas ${getActionIcon(action.action_type)}"></i>
          ${action.action_type}: ${action.emp_name}
        </div>
        <div>${action.details}</div>
        <div class="action-date">
          <i class="far fa-clock"></i> ${new Date(action.action_date).toLocaleString()}
        </div>
      </div>
    `;
    
    // Add new action at the top with animation
    actionsContainer.insertAdjacentHTML('afterbegin', actionHTML);
    
    // Highlight the new action temporarily
    const newAction = actionsContainer.firstElementChild;
    setTimeout(() => newAction.classList.remove('new-action'), 2000);
  }
  
  function showToast(message, type = 'info') {
    const toast = document.createElement('div');
    toast.className = `toast toast-${type}`;
    toast.innerHTML = `
      <i class="fas ${getToastIcon(type)}"></i>
      <span>${message}</span>
    `;
    
    document.body.appendChild(toast);
    
    // Auto-remove after delay
    setTimeout(() => {
      toast.classList.add('fade-out');
      setTimeout(() => toast.remove(), 500);
    }, 5000);
  }
  
  function getToastIcon(type) {
    const icons = {
      'success': 'fa-check-circle',
      'error': 'fa-exclamation-circle',
      'warning': 'fa-exclamation-triangle',
      'info': 'fa-info-circle'
    };
    return icons[type] || 'fa-info-circle';
  }
  
  // Close popup functions
  function closeSuccessPopup() {
    const popup = document.querySelector('.success-popup');
    popup.classList.add('fade-out');
    setTimeout(() => popup.remove(), 500);
  }
  
  function closeFirePopup() {
    const popup = document.querySelector('.fire-popup');
    popup.classList.add('fade-out');
    setTimeout(() => popup.remove(), 500);
  }
  
  // Logout function
  async function logout() {
    try {
      const response = await fetch(`${CONFIG.apiBaseUrl}/logout`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        credentials: 'include'
      });
  
      const result = await response.json();
  
      if (result.success) {
        // Clear session storage
        sessionStorage.clear();
        // Redirect to login page
        window.location.href = '/login.html';
      } else {
        throw new Error(result.error || 'Logout failed');
      }
    } catch (error) {
      console.error('Logout error:', error);
      showToast('Failed to logout. Please try again.', 'error');
    }
  }

File: hr.py
from flask import Flask, request, jsonify, session, redirect, url_for, send_from_directory
from flask_cors import CORS
import mysql.connector
import bcrypt
import random
import string
from datetime import datetime
import logging

app = Flask(__name__, static_folder='templates')
app.secret_key = 'your_secure_secret_key_here_12345!'
app.config['SESSION_COOKIE_SAMESITE'] = 'Lax'
app.config['SESSION_COOKIE_SECURE'] = False  # Set to True in production with HTTPS
app.config['SESSION_COOKIE_HTTPONLY'] = True
app.config['PERMANENT_SESSION_LIFETIME'] = 3600  # 1 hour session lifetime

# Configure logging
logging.basicConfig(level=logging.DEBUG)
logger = logging.getLogger(__name__)

CORS(app, supports_credentials=True, resources={
    r"/api/*": {
        "origins": ["http://localhost:3000", "http://127.0.0.1:3000"],
        "methods": ["GET", "POST", "PUT", "DELETE"],
        "allow_headers": ["Content-Type", "Authorization"]
    }
})

# Database configuration
DB_CONFIG = {
    'host': 'localhost',
    'user': 'abel tiruneh',
    'password': 'Ab1996@2468',
    'database': 'TIME_BANK',
    'auth_plugin': 'mysql_native_password'  # Add if using MySQL 8+
}

# Department mapping
DEPARTMENT_MAPPING = {
    'Accountant': 101,
    'Manager': 102,
    'Finance': 103,
    'Security': 104,
    'Cleaner': 105,
    'HR': 107
}

def get_db():
    try:
        conn = mysql.connector.connect(**DB_CONFIG)
        logger.info("Database connection successful")
        return conn
    except Exception as e:
        logger.error(f"Database connection failed: {e}")
        return None

def generate_employee_id():
    return random.randint(1000, 9999)

def generate_password(length=8):
    chars = string.ascii_letters + string.digits
    return ''.join(random.choice(chars) for _ in range(length))

@app.route('/api/login', methods=['POST'])
def login():
    logger.info("Login attempt received")
    data = request.get_json()
    username = data.get('username')
    password = data.get('password')

    if not username or not password:
        logger.warning("Missing username or password")
        return jsonify({'success': False, 'error': 'Username and password are required'}), 400

    db = get_db()
    if not db:
        logger.error("Database connection failed during login")
        return jsonify({'success': False, 'error': 'Database connection failed'}), 500

    try:
        cursor = db.cursor(dictionary=True)
        cursor.execute("""
            SELECT e.*, d.dep_name 
            FROM employee e
            JOIN department d ON e.dep_id = d.dep_id
            WHERE username = %s AND job_title IS NOT NULL
        """, (username,))
        user = cursor.fetchone()

        if not user:
            logger.warning(f"No active user found with username: {username}")
            return jsonify({'success': False, 'error': 'Invalid username or password'}), 401

        logger.info(f"User found: {user['emp_name']}")
        logger.debug(f"Password comparison: {password} vs stored hash")

        # Verify password
        if bcrypt.checkpw(password.encode('utf-8'), user['passwords'].encode('utf-8')):
            logger.info("Password matches - login successful")
            
            # Store user info in session
            session.clear()
            session['user_id'] = user['emp_id']
            session['username'] = user['username']
            session['name'] = user['emp_name']
            session['emp_id'] = user['emp_id']
            session['job_title'] = user['job_title']
            session['department'] = user['dep_name']
            session.permanent = True
            
            logger.debug(f"Session data: {dict(session)}")
            
            return jsonify({
                'success': True,
                'user': {
                    'emp_id': user['emp_id'],
                    'name': user['emp_name'],
                    'job_title': user['job_title'],
                    'department': user['dep_name']
                }
            })
        else:
            logger.warning("Password does not match")
            return jsonify({'success': False, 'error': 'Invalid username or password'}), 401

    except Exception as e:
        logger.error(f"Login error: {str(e)}", exc_info=True)
        return jsonify({'success': False, 'error': 'Login failed'}), 500
    finally:
        cursor.close()
        db.close()

@app.route('/api/logout', methods=['POST'])
def logout():
    logger.info(f"Logout requested by user {session.get('username')}")
    session.clear()
    return jsonify({'success': True, 'message': 'Logged out successfully'})

@app.route('/api/current-user')
def current_user():
    if 'user_id' not in session:
        logger.warning("Current user request without valid session")
        return jsonify({'success': False, 'error': 'Not authenticated'}), 401

    logger.debug(f"Current user session: {dict(session)}")
    return jsonify({
        'success': True,
        'user': {
            'emp_id': session.get('emp_id'),
            'name': session.get('name'),
            'username': session.get('username'),
            'job_title': session.get('job_title'),
            'role': session.get('department')
        }
    })

@app.route('/api/department-stats')
def department_stats():
    if 'user_id' not in session:
        return jsonify({'success': False, 'error': 'Not authenticated'}), 401

    db = get_db()
    if not db:
        return jsonify({'success': False, 'error': 'Database connection failed'}), 500

    try:
        cursor = db.cursor(dictionary=True)
        cursor.execute("""
            SELECT d.dep_name, COUNT(e.emp_id) as employee_count
            FROM department d
            LEFT JOIN employee e ON d.dep_id = e.dep_id AND e.job_title IS NOT NULL
            GROUP BY d.dep_name
            ORDER BY d.dep_name
        """)
        stats = cursor.fetchall()
        return jsonify({'success': True, 'stats': stats})
    except Exception as e:
        logger.error(f"Department stats error: {e}")
        return jsonify({'success': False, 'error': 'Failed to fetch department stats'}), 500
    finally:
        cursor.close()
        db.close()

@app.route('/api/recent-actions')
def recent_actions():
    if 'user_id' not in session:
        return jsonify({'success': False, 'error': 'Not authenticated'}), 401

    db = get_db()
    if not db:
        return jsonify({'success': False, 'error': 'Database connection failed'}), 500

    try:
        cursor = db.cursor(dictionary=True)
        cursor.execute("""
            SELECT ea.*, e.emp_name 
            FROM employee_actions ea
            JOIN employee e ON ea.emp_id = e.emp_id
            ORDER BY ea.action_date DESC
            LIMIT 10
        """)
        actions = cursor.fetchall()
        return jsonify({'success': True, 'actions': actions})
    except Exception as e:
        logger.error(f"Recent actions error: {e}")
        return jsonify({'success': False, 'error': 'Failed to fetch recent actions'}), 500
    finally:
        cursor.close()
        db.close()

@app.route('/api/hire', methods=['POST'])
def hire_employee():
    if 'user_id' not in session:
        return jsonify({'success': False, 'error': 'Not authenticated'}), 401

    data = request.get_json()
    required_fields = ['name', 'phone', 'email', 'city', 'address', 'role', 'branch_id', 'dob', 'gender', 'salary']
    if not all(field in data for field in required_fields):
        return jsonify({'success': False, 'error': 'Missing required fields'}), 400

    db = get_db()
    if not db:
        return jsonify({'success': False, 'error': 'Database connection failed'}), 500

    try:
        cursor = db.cursor()

        # Generate employee ID
        emp_id = generate_employee_id()
        
        # Generate username (firstname.lastname)
        first_name = data['name'].split()[0].lower()
        last_name = data['name'].split()[-1].lower() if len(data['name'].split()) > 1 else ''
        username = f"{first_name}.{last_name}" if last_name else first_name
        
        # Check if username exists and append number if needed
        temp_username = username
        counter = 1
        while True:
            cursor.execute("SELECT emp_id FROM employee WHERE username = %s", (temp_username,))
            if not cursor.fetchone():
                username = temp_username
                break
            temp_username = f"{username}{counter}"
            counter += 1

        # Generate temporary password
        temp_password = generate_password()
        hashed_password = bcrypt.hashpw(temp_password.encode('utf-8'), bcrypt.gensalt()).decode('utf-8')

        # Get department ID
        dep_id = DEPARTMENT_MAPPING.get(data['role'])
        if not dep_id:
            return jsonify({'success': False, 'error': 'Invalid role'}), 400

        # Insert new employee
        cursor.execute("""
            INSERT INTO employee (
                emp_id, emp_name, gender, dep_id, branch_id, job_title, salary, dbo, 
                phone, city, address, email, username, passwords
            ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        """, (
            emp_id, data['name'], data['gender'], dep_id, data['branch_id'], data['role'], 
            data['salary'], data['dob'], data['phone'], data['city'], data['address'], 
            data['email'], username, hashed_password
        ))

        # Record the hiring action
        cursor.execute("""
            INSERT INTO employee_actions (emp_id, action_type, details)
            VALUES (%s, 'Hire', %s)
        """, (emp_id, f"Hired as {data['role']} with salary {data['salary']}"))

        db.commit()

        logger.info(f"New employee hired: {emp_id} - {data['name']}")

        return jsonify({
            'success': True,
            'emp_id': emp_id,
            'username': username,
            'temp_password': temp_password
        })

    except Exception as e:
        db.rollback()
        logger.error(f"Hire employee error: {e}")
        return jsonify({'success': False, 'error': 'Failed to hire employee'}), 500
    finally:
        cursor.close()
        db.close()

@app.route('/api/fire', methods=['POST'])
def fire_employee():
    if 'user_id' not in session:
        return jsonify({'success': False, 'error': 'Not authenticated'}), 401

    data = request.get_json()
    if not data.get('empId') or not data.get('name'):
        return jsonify({'success': False, 'error': 'Employee ID and name are required'}), 400

    db = get_db()
    if not db:
        return jsonify({'success': False, 'error': 'Database connection failed'}), 500

    try:
        cursor = db.cursor(dictionary=True)

        # Verify employee exists
        cursor.execute("SELECT * FROM employee WHERE emp_id = %s AND emp_name = %s AND job_title IS NOT NULL", 
                      (data['empId'], data['name']))
        employee = cursor.fetchone()
        
        if not employee:
            return jsonify({'success': False, 'error': 'Employee not found or already terminated'}), 404

        # Record the firing action before deleting
        cursor.execute("""
            INSERT INTO employee_actions (emp_id, action_type, details)
            VALUES (%s, 'Fire', 'Employee terminated')
        """, (data['empId'],))

        # Remove employee (set job_title to NULL as per your trigger)
        cursor.execute("""
            UPDATE employee 
            SET job_title = NULL 
            WHERE emp_id = %s
        """, (data['empId'],))

        db.commit()

        logger.info(f"Employee terminated: {data['empId']} - {data['name']}")

        return jsonify({'success': True})

    except Exception as e:
        db.rollback()
        logger.error(f"Fire employee error: {e}")
        return jsonify({'success': False, 'error': 'Failed to terminate employee'}), 500
    finally:
        cursor.close()
        db.close()

@app.route('/api/update-credentials', methods=['POST'])
def update_credentials():
    if 'user_id' not in session:
        return jsonify({'success': False, 'error': 'Not authenticated'}), 401

    data = request.get_json()
    required_fields = ['currentUsername', 'newUsername', 'currentPassword', 'newPassword']
    if not all(field in data for field in required_fields):
        return jsonify({'success': False, 'error': 'All fields are required'}), 400

    db = get_db()
    if not db:
        return jsonify({'success': False, 'error': 'Database connection failed'}), 500

    try:
        cursor = db.cursor(dictionary=True)

        # Verify current credentials
        cursor.execute("""
            SELECT passwords FROM employee 
            WHERE emp_id = %s AND username = %s
        """, (session['user_id'], data['currentUsername']))
        user = cursor.fetchone()

        if not user:
            return jsonify({'success': False, 'error': 'Current username is incorrect'}), 401

        if not bcrypt.checkpw(data['currentPassword'].encode('utf-8'), user['passwords'].encode('utf-8')):
            return jsonify({'success': False, 'error': 'Current password is incorrect'}), 401

        # Check if new username is already taken
        if data['newUsername'] != data['currentUsername']:
            cursor.execute("SELECT emp_id FROM employee WHERE username = %s", (data['newUsername'],))
            if cursor.fetchone():
                return jsonify({'success': False, 'error': 'New username is already taken'}), 400

        # Hash new password
        hashed_password = bcrypt.hashpw(data['newPassword'].encode('utf-8'), bcrypt.gensalt()).decode('utf-8')

        # Update credentials
        cursor.execute("""
            UPDATE employee 
            SET username = %s, passwords = %s 
            WHERE emp_id = %s
        """, (data['newUsername'], hashed_password, session['user_id']))

        db.commit()

        # Update session
        session['username'] = data['newUsername']

        logger.info(f"Credentials updated for user: {session['user_id']}")

        return jsonify({'success': True})

    except Exception as e:
        db.rollback()
        logger.error(f"Update credentials error: {e}")
        return jsonify({'success': False, 'error': 'Failed to update credentials'}), 500
    finally:
        cursor.close()
        db.close()

@app.route('/')
def home():
    return redirect(url_for('login_page'))

@app.route('/login')
def login_page():
    return send_from_directory('templates', 'login.html')

@app.route('/hr-dashboard')
def hr_dashboard():
    if 'user_id' not in session:
        return redirect(url_for('login_page'))
    return send_from_directory('templates', 'HR-Dashboard.html')

# Serve static files
@app.route('/static/<path:path>')
def serve_static(path):
    return send_from_directory('static', path)

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=8000, debug=True)